packages/lib/src/components/internal/UIElement/UIElement.tsx

```diff
@@ -190,12 +199,14 @@ export abstract class UIElement<P extends UIElementProps = UIElementProps> exten
                 ? new Promise((resolve, reject) =>
                       this.props.beforeSubmit(this.data, this.elementRef, {
                           resolve,
-                          reject
+                          reject: () => reject('beforeSubmitRejected')
                       })
                   )
                 : Promise.resolve(this.data);
 
-            return beforeSubmitEvent.then(this.submitUsingSessionsFlow);
+            return beforeSubmitEvent.then(this.submitUsingSessionsFlow).catch(e => {
+                throw new Error(e);
+            });
```

Comment:
We think it can be simplified a bit by handling the error only at the `this.makePaymentsCall` catch block.
One way would be to reject by passing an custom `CancelError` object in `beforeSubmit` promise:

```js
const beforeSubmitEvent: Promise<PaymentData> = this.props.beforeSubmit
            ? new Promise((resolve, reject) =>
                  this.props.beforeSubmit(this.data, this.elementRef, {
                      resolve,
                      reject: () => reject(new CancelError()),
                  })
              )
            : Promise.resolve(this.data);

return beforeSubmitEvent.then(this.submitUsingSessionsFlow);
```

Then when handling the rejection, we could do this way:

```js
this.makePaymentsCall()
            .then(sanitizeResponse)
            .then(verifyPaymentDidNotFail)
            .then(this.handleResponse)
            .catch((error: PaymentResponseData | CancelError) => {
                if (error instanceof CancelError) {
                    this.setElementStatus('ready');
                    return;
                }
                
                this.handleFailedResult(error);
            });
```

This `CancelError` is just an internal error object that we can have on our codebase to cancel certain operations internally. In this case, we use for the `beforeSubmit` , but in the future we can use for other rejections too (Ex: if we have eventually a `beforeAdditionalDetails` , or any other callbacks). 

It can be a simple error object like this:

```js
class CancelError extends Error {}
```

So we can easily extend it in the future if needed. No need to pass any parameters to it for now.